src/tests/t_kdb.py

```diff
@@ -336,6 +336,31 @@ def ldap_add(dn, objectclass, attrs=[]):
 realm.kinit('canon', password('canon'))
 realm.kinit('alias', password('canon'), ['-C'])
 
+# Test password history
+def test_pwhist(nhist):
+    output('Testing password history of size ' + `nhist`)
+    princ = 'pwhistprinc' + `nhist`
+    pol = 'pwhistpol'+`nhist`
+    realm.run([kadminl, 'addpol', '-history', `nhist`, pol])
+    realm.run([kadminl, 'addprinc', '-policy', pol, '-nokey', princ])
+    # Test nhist number of passwords and make sure that all previous passwords
+    # fail when trying to be changed to them.
+    cpw = lambda x,y: realm.run([kadminl, 'cpw', '-pw', `x`, princ],
+                                expected_code=y)
+
+    # Creates a combination of password attempts, with (x, 0) and ([0..x], 1)
+    cmb = lambda x: izip(chain([x], islice(count(), x+1)),
+                             chain([0], repeat(1)))
+    # Tries passwords, with one password change that should succeed followed by
+    # all previous passwords that should fail
+    # (0, 0) (0, 1) (1, 0) (0, 1), (1, 1)...
+    list(starmap(cpw, chain.from_iterable(imap(cmb, islice(count(), nhist)))))
+
+    # Test that keys are rotated out correctly
+    cpw(nhist, 0)
+    cpw(0, 0)
+map(test_pwhist, imap(pow, repeat(2), range(5)))
```

Comment:
Testing a history size of 16 isn't likely to expose any more problems than testing a history size of 3; there are no additional edge cases.  Running hundreds of commands when a dozen would suffice makes it more difficult to sort through the testing sequence and debug problems when a test in t_kdb.py fails.  And while the ten extra seconds of runtime might not be a problem in isolation, when multiplied by the hundreds of similarly-sized pieces of functionality in krb5, it could make the test suite take so long that developers will stop wanting to run it before submitting changes.

Most Python resources I can find discourage using lambda for non-anonymous functions (use def instead), encourage using list comprehensions and generator expressions over using map and filter, and discourage using the backtick operator (which I had actually never heard of before).  For instance:

  http://stackoverflow.com/questions/134626/which-is-more-preferable-to-use-in-python-lambda-functions-or-nested-functions
  http://stackoverflow.com/questions/1673071/what-do-backticks-mean-to-the-python-interpreter-num

In this case, I think the edge cases can be hit with a history size of 2 or 3 and no iteration or functions in the Python code, just a straightforward list of commands with expected outcomes.  If any of the commands fail, it's much easier for a developer to figure out where the failure is in the test script and what was supposed to happen.